### DISCUSSION TITLE
Default mutation function using key

### DISCUSSION THREAD

**Discussion by honzahruby** [TanStack/query#7442] [Ideas] [answered]
Similar to the [default query function](https://tanstack.com/query/latest/docs/framework/react/guides/default-query-function#default-query-function), we'd love the ability for the default mutation function to access the mutation key. Currently, it can only access the variables passed by the user.

Is it a conscious decision, given the mutation key is optional, while in queries it's required? Or is it something to consider adding? 🙏

**Comment by honzahruby** [comment]
@TkDodo any idea why it's so? 🙏

**Comment by TkDodo** [comment]
we don't pass anything to the mutationFn because the mutationFn takes `variables` passed in by users already.

**Reply by honzahruby** [reply]
I know it's not possible now, but how about passing some extra context as a second argument then? 

Using the default query function is great because we don't need to generate any extra code for our API client, but it's unfortunate the same doesn't work for mutations 😕

**Comment by honzahruby** [comment]
Sorry for pinging you again @TkDodo, I've put together a simple code example to better explain what I'm trying to achieve. 

Basically going from this (where the key is optional)
```ts
const useDeleteUser = () =>
    useMutation({
        mutationKey: ['users', 'delete'],
        mutationFn: async (id: string) => apiClient.users.delete(id),
    });
```
to this
```ts
const reactQueryClient = new QueryClient({
    defaultOptions: {
        mutations: {
            mutationFn: async (variables, { mutationKey: [controller, method] }) => apiClient[controller][method](variables),
        },
    },
});

const useDeleteUser = () =>
    useMutation({
        mutationKey: ['users', 'delete'],
    });
```
which mimics what's already possible for queries.

Is it something you would consider adding? 🙏

**Reply by TkDodo** [reply]
I'll think about it, yes

**Reply by honzahruby** [reply, marked as answer]
Closing as this has been implemented in #9615. Huge thank you to @TkDodo, @joshuaellis and others involved 👏
